software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+    vma_region   -- name of the VMA region if the section is to be directed to
+                    a VMA region, else None.
+    lma_region   -- name of the LMA region if the section is to be directed to
+                    a VMA region, else None.
+    in_sections  -- list of input sections.
+    in_objects   -- list of input objects to consider.
+    """
+    vaddr        = "" # virtual address
+    laddr        = "" # logical address
+    vma_redirect = ""
+    lma_redirect = ""
+
+    if lma_region == None:
+      laddr = "AT({0})".format(address)
+      vma_redirect = ">{0}".format(vma_region)
+    else:
+      vaddr = address
+      lma_redirect = "AT>{0}".format(lma_region)
+
+    script = "{0} {1} :\n{2} {{".format(name, vaddr, laddr)
+
+    for sec in in_sections:
+      script += "\n  {0}({1})".format(in_objects, sec)
+
+    script += "\n  . = ALIGN(8);\n"
+    script += "\n}} {0} {1}\n\n".format(vma_redirect, lma_redirect)
+
+    return script
+
+  def script(self):
+    # Address constants:
+    #
+    # LMA (Load Memory Address)    => NPA used by loader
+    # VMA (Virtual Memory Address) => Logical address used by linker for 
+    #                                 symbol resolutions
+    _DMEM_VMA_START = 0x1000
+    _DMEM_VMA_SIZE  = 0x1000
+    _DRAM_LMA_START  = 0x80000000
+    _DRAM_LMA_SIZE   = self._dram_size
+
+    mem_regions = [
+      # Format:
+      # [<mem region name>, <access>, <start address>, <size>]
+      #
+      # .text section VMAs and dmem memory VMAs overlap with the
+      # current memory layout (both start at address 0x0). This makes
+      # manycore essentially a Harvard architecture. But linker doesn't
+      # allow overlapping VMAs by not letting location couter not move 
+      # backwards. To overcome this, we use combination of VMA and LMA
+      # "memory regions" to be able to never move location counter
+      # backward.
+      ['DMEM_VMA', 'rw', _DMEM_VMA_START, _DMEM_VMA_SIZE],
+      ['DRAM_LMA', 'rw', _DRAM_LMA_START, _DRAM_LMA_SIZE],
+      ]
+
+    section_map = [
+      # Format:
+      # <output section>: [<input sections>]
+      ['.text.dram'        , ['.crtbegin','.text','.text.startup','.text.*']],
+      ['.dmem'             , ['.dmem','.dmem.*']],
+      ['.data'             , ['.data','.data*']],
+      ['.sdata'            , ['.srodata.cst16','.srodata.cst8','.srodata.cst4',
+                              '.srodata.cst2','.srodata*','.sdata','.sdata.*',
+                              '.gnu.linkonce.s.*']],
+      ['.sbss'             , ['.sbss','.sbss.*','.gnu.linkonce.sb.*','.scommon']],
+      ['.bss'              , ['.bss','.bss*']],
+      ['.tdata'            , ['.tdata','.tdata*']],
+      ['.tbss'             , ['.tbss','.tbss*']],
+      ['.eh_frame'         , ['.eh_frame','.eh_frame*']],
+      ['.striped.data.dmem', ['.striped.data']],
+      ['.rodata.dram'      , ['.rodata','.rodata*']],
+      ['.dram'             , ['.dram','.dram.*', '.*']],
+      ]
```

Comment:
Do you want this to be a list of lists, or a dictionary with list values? (My sense is the former, because ordering seems to matter)

Reply:
Yep, natural choice are dicts, but realized order matters here. For eg., `.crtbegin` section should always be at the top...